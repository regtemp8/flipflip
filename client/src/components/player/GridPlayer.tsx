import React, {
  useCallback,
  useEffect,
  useState,
  useRef,
  type ReactNode
} from 'react'
import { cx } from '@emotion/css'

import {
  AppBar,
  Container,
  IconButton,
  type Theme,
  Toolbar,
  Tooltip,
  Typography
} from '@mui/material'

import { makeStyles } from 'tss-react/mui'

import ArrowBackIcon from '@mui/icons-material/ArrowBack'
import FullscreenIcon from '@mui/icons-material/Fullscreen'

import Player from './Player'
import type ChildCallbackHack from './ChildCallbackHack'
import { IdleTimer } from './IdleTimer'
import { useAppDispatch, useAppSelector } from '../../store/hooks'
import { setRouteGoBack } from '../../store/app/thunks'
import {
  selectSceneGridCellMatchesSceneCopy,
  selectSceneGridCellMirror,
  selectSceneGridCellSceneCopy,
  selectSceneGridCellSceneID,
  selectSceneGridName
} from '../../store/sceneGrid/selectors'
import {
  selectGridPlayerFirstNotLoaded,
  selectPlayerOverlayGrid,
  selectPlayerOverlaySceneID
} from '../../store/player/selectors'
import flipflip from '../../FlipFlipService'
import { setFullScreen, toggleFullScreen } from '../../data/actions'

const useStyles = makeStyles()((theme: Theme) => ({
  root: {
    display: 'flex',
    position: 'fixed',
    top: 0,
    right: 0,
    bottom: 0,
    left: 0
  },
  appBar: {
    zIndex: theme.zIndex.drawer + 1,
    height: theme.spacing(8),
    marginTop: -theme.spacing(8) - 3,
    transition: theme.transitions.create('margin', {
      easing: theme.transitions.easing.sharp,
      duration: theme.transitions.duration.leavingScreen
    })
  },
  appBarHover: {
    marginTop: 0,
    transition: theme.transitions.create('margin', {
      easing: theme.transitions.easing.sharp,
      duration: theme.transitions.duration.enteringScreen
    })
  },
  hoverBar: {
    zIndex: theme.zIndex.drawer + 1,
    position: 'absolute',
    opacity: 0,
    height: theme.spacing(5),
    width: '100%',
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'flex-end',
    padding: '0 8px',
    minHeight: 64
  },
  title: {
    textAlign: 'center'
  },
  content: {
    display: 'flex',
    flexGrow: 1,
    flexDirection: 'column',
    height: '100vh'
  },
  container: {
    height: '100%',
    padding: theme.spacing(0)
  },
  grid: {
    flexGrow: 1,
    display: 'grid',
    height: '100%'
  },
  gridCell: {
    height: '100%',
    width: '100%',
    display: 'grid',
    overflow: 'hidden'
  },
  fill: {
    flexGrow: 1
  },
  hidden: {
    opacity: 0
  },
  hideCursor: {
    cursor: 'none'
  },
  mirror: {
    transform: 'scaleX(-1)'
  }
}))

interface GridCellPlayerProps {
  gridID: number
  row: number
  col: number
  width: number
  height: number
  uuid: string
  advanceHack?: any
  allLoaded: boolean
  sceneCopyGrid: ReactNode[][]

  setProgress?: (total: number, current: number, message: string[]) => void
  setSceneCopy: (row: number, col: number, children: ReactNode) => void
  setVideo?: (index: number, video: HTMLVideoElement) => void
}

function GridCellPlayer(props: GridCellPlayerProps) {
  const { classes } = useStyles()
  const { gridID, row, col, width, height, uuid, allLoaded } = props

  const sceneID = useAppSelector(selectSceneGridCellSceneID(gridID, row, col))
  const sceneCopy = useAppSelector(
    selectSceneGridCellSceneCopy(gridID, row, col)
  )
  const mirror = useAppSelector(selectSceneGridCellMirror(gridID, row, col))
  const matchesSceneCopy = useAppSelector(
    selectSceneGridCellMatchesSceneCopy(gridID, [row, col])
  )

  if (sceneCopy.length > 0) {
    const sceneCopyGridCell = props.sceneCopyGrid[sceneCopy[0]][sceneCopy[1]]
    return (
      <div
        className={cx(
          classes.gridCell,
          !sceneCopyGridCell && classes.hidden,
          mirror && classes.mirror
        )}
      >
        <div
          style={{
            left: 0,
            right: 0,
            top: 0,
            bottom: 0,
            display: 'flex',
            overflow: 'hidden',
            position: 'relative'
          }}
        >
          <div>
            <div
              style={{
                position: 'absolute',
                top: 0,
                bottom: 0,
                left: 0,
                right: 0
              }}
            >
              <div
                style={{
                  top: 0,
                  right: 0,
                  bottom: 0,
                  left: 0,
                  position: 'static'
                }}
              >
                {sceneCopyGridCell}
              </div>
            </div>
          </div>
        </div>
      </div>
    )
  } else {
    return (
      <div className={cx(classes.gridCell, !sceneID && classes.hidden)}>
        {sceneID && (
          <Player
            uuid={uuid}
            preventSleep={row === 0 && col === 0}
            advanceHack={props.advanceHack}
            gridView
            gridCoordinates={matchesSceneCopy ? [row, col] : undefined}
            captionScale={1 / Math.sqrt(width * height)}
            allLoaded={allLoaded}
            setProgress={props.setProgress}
            setSceneCopy={(children: ReactNode) =>
              props.setSceneCopy(row, col, children)
            }
            setVideo={(video: HTMLVideoElement) => {
              if (props.setVideo != null) {
                props.setVideo(row * width + col, video)
              }
            }}
          />
        )}
      </div>
    )
  }
}

export interface GridPlayerProps {
  parentUUID: string
  overlayIndex: number
  advanceHacks?: ChildCallbackHack[]
  hideBars?: boolean
  setProgress?: (total: number, current: number, message: string[]) => void
  setVideo?: (index: number, video: HTMLVideoElement) => void
}

function GridPlayer(props: GridPlayerProps) {
  const dispatch = useAppDispatch()
  const firstNotLoaded = useAppSelector(
    selectGridPlayerFirstNotLoaded(props.overlayIndex, props.parentUUID)
  )
  const allLoaded = firstNotLoaded[0] === -1 && firstNotLoaded[1] === -1
  const sceneID = useAppSelector(
    selectPlayerOverlaySceneID(props.overlayIndex, props.parentUUID)
  )
  const grid = useAppSelector(
    selectPlayerOverlayGrid(props.overlayIndex, props.parentUUID)
  )
  const name = useAppSelector(selectSceneGridName(sceneID))

  const [appBarHover, setAppBarHover] = useState(false)
  const [hideCursor, setHideCursor] = useState(false)
  const [sceneCopyGrid, setSceneCopyGrid] = useState(
    grid.map((r) => r.map((c) => null)) as React.ReactNode[][]
  )

  const _idleTimerRef = useRef<any>()
  const _appBarTimeout = useRef<any>()

  const navigateBack = useCallback(async () => {
    setFullScreen(false)
    dispatch(setRouteGoBack())
  }, [dispatch])

  useEffect(() => {
    const abortController = new AbortController()
    flipflip().events.onGridNavigateBack(navigateBack, abortController)

    return () => {
      abortController.abort()
      if (_appBarTimeout.current) {
        clearTimeout(_appBarTimeout.current)
      }
    }
  }, [navigateBack])

  const setSceneCopy = (
    rowIndex: number,
    colIndex: number,
    children: React.ReactNode
  ) => {
    const newSceneCopyGrid = sceneCopyGrid
    newSceneCopyGrid[rowIndex][colIndex] = children
    setSceneCopyGrid(newSceneCopyGrid)
  }

  const onActive = () => {
    setHideCursor(false)
  }

  const onIdle = () => {
    setHideCursor(true)
  }

  const onMouseEnterAppBar = () => {
    clearTimeout(_appBarTimeout.current)
    setAppBarHover(true)
  }

  const closeAppBar = () => {
    setAppBarHover(false)
  }

  const onMouseLeaveAppBar = () => {
    clearTimeout(_appBarTimeout.current)
    _appBarTimeout.current = window.setTimeout(closeAppBar, 1000)
  }

  const { classes } = useStyles()
  const height = grid && grid.length > 0 && grid[0].length > 0 ? grid.length : 1
  const width =
    grid && grid.length > 0 && grid[0].length > 0 ? grid[0].length : 1

  const colSize = 100 / width
  const rowSize = 100 / height
  let gridTemplateColumns = ''
  let gridTemplateRows = ''
  for (let w = 0; w < width; w++) {
    gridTemplateColumns += colSize.toString() + '% '
  }
  for (let h = 0; h < height; h++) {
    gridTemplateRows += rowSize.toString() + '% '
  }

  return (
    <div className={classes.root}>
      {!props.hideBars && (
        <React.Fragment>
          <div
            className={classes.hoverBar}
            onMouseEnter={onMouseEnterAppBar}
            onMouseLeave={onMouseLeaveAppBar}
          />

          <AppBar
            enableColorOnDark
            position="absolute"
            onMouseEnter={onMouseEnterAppBar}
            onMouseLeave={onMouseLeaveAppBar}
            className={cx(classes.appBar, appBarHover && classes.appBarHover)}
          >
            <Toolbar>
              <Tooltip disableInteractive title="Back" placement="right-end">
                <IconButton
                  edge="start"
                  color="inherit"
                  aria-label="Back"
                  onClick={() => {
                    dispatch(setRouteGoBack())
                  }}
                  size="large"
                >
                  <ArrowBackIcon />
                </IconButton>
              </Tooltip>

              <div className={classes.fill} />
              <Typography
                component="h1"
                variant="h4"
                color="inherit"
                noWrap
                className={classes.title}
              >
                {name}
              </Typography>
              <div className={classes.fill} />

              <Tooltip disableInteractive title="Toggle Fullscreen">
                <IconButton
                  edge="start"
                  color="inherit"
                  aria-label="FullScreen"
                  onClick={toggleFullScreen}
                  size="large"
                >
                  <FullscreenIcon fontSize="large" />
                </IconButton>
              </Tooltip>
            </Toolbar>
          </AppBar>
        </React.Fragment>
      )}

      <main
        className={cx(classes.content, hideCursor && classes.hideCursor)}
        ref={_idleTimerRef}
      >
        <IdleTimer
          ref={_idleTimerRef}
          onActive={onActive}
          onIdle={onIdle}
          timeout={2000}
        />
        <Container maxWidth={false} className={classes.container}>
          <div
            className={classes.grid}
            style={{
              gridTemplateColumns,
              gridTemplateRows
            }}
          >
            {[...Array(height).keys()].map((row) => (
              <React.Fragment key={row}>
                {[...Array(width).keys()].map((col) => {
                  const uuid = grid[row][col]
                  const showProgress =
                    firstNotLoaded[0] === row && firstNotLoaded[1] === col
                  return (
                    <GridCellPlayer
                      key={col}
                      gridID={sceneID}
                      row={row}
                      col={col}
                      width={width}
                      height={height}
                      uuid={uuid}
                      advanceHack={
                        props.advanceHacks
                          ? props.advanceHacks[row * width + col]
                          : undefined
                      }
                      allLoaded={allLoaded}
                      sceneCopyGrid={sceneCopyGrid}
                      setVideo={props.setVideo}
                      setProgress={showProgress ? props.setProgress : undefined}
                      setSceneCopy={setSceneCopy}
                    />
                  )
                })}
              </React.Fragment>
            ))}
          </div>
        </Container>
      </main>
    </div>
  )
}

;(GridPlayer as any).displayName = 'GridPlayer'
export default GridPlayer
