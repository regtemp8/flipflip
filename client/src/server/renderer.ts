import createCache, { type EmotionCache } from '@emotion/cache'
import { configureStore } from '@reduxjs/toolkit'
import { AppStorage, SystemConstants } from 'flipflip-common'
import createAppReducer from '../store/app/slice'
import createAudioReducer from '../store/audio/slice'
import createCaptionScriptReducer from '../store/captionScript/slice'
import createClipReducer from '../store/clip/slice'
import createLibrarySourceReducer from '../store/librarySource/slice'
import createPlaylistReducer from '../store/playlist/slice'
import createSceneReducer from '../store/scene/slice'
import createSceneGroupReducer from '../store/sceneGroup/slice'
import createTagReducer from '../store/tag/slice'
import playersSlice from '../store/player/slice'
import videoClipperSlice from '../store/videoClipper/slice'
import scenePickerSlice from '../store/scenePicker/slice'
import captionScriptorSlice from '../store/captionScriptor/slice'
import sceneDetailSlice from '../store/sceneDetail/slice'
import audioLibrarySlice from '../store/audioLibrary/slice'
import sourceScraperSlice from '../store/sourceScraper/slice'
import createConstantsReducer from '../store/constants/slice'
import { fromAppStorage } from '../store/app/convert'
import createDisplayReducer from '../store/display/slice'
import createDisplayViewReducer from '../store/displayView/slice'
import createScenePlaylistItemReducer from '../store/scenePlaylistItem/slice'
import createDisplayPlaylistItemReducer from '../store/displayPlaylistItem/slice'

export function createEmotionCache(): EmotionCache {
  return createCache({ key: 'css' })
}

export function createReduxStore(
  appStorage?: AppStorage,
  constants?: SystemConstants
) {
  const state = appStorage != null ? fromAppStorage(appStorage) : undefined
  return configureStore({
    reducer: {
      app: createAppReducer(state?.app),
      audio: createAudioReducer(state?.audio),
      captionScript: createCaptionScriptReducer(state?.captionScript),
      clip: createClipReducer(state?.clip),
      librarySource: createLibrarySourceReducer(state?.librarySource),
      playlist: createPlaylistReducer(state?.playlist),
      displayPlaylistItem: createDisplayPlaylistItemReducer(
        state?.displayPlaylistItem
      ),
      scenePlaylistItem: createScenePlaylistItemReducer(
        state?.scenePlaylistItem
      ),
      scene: createSceneReducer(state?.scene),
      sceneGroup: createSceneGroupReducer(state?.sceneGroup),
      display: createDisplayReducer(state?.display),
      displayView: createDisplayViewReducer(state?.displayView),
      tag: createTagReducer(state?.tag),
      constants: createConstantsReducer(constants),
      // components
      players: playersSlice,
      videoClipper: videoClipperSlice,
      scenePicker: scenePickerSlice,
      captionScriptor: captionScriptorSlice,
      sceneDetail: sceneDetailSlice,
      audioLibrary: audioLibrarySlice,
      sourceScraper: sourceScraperSlice
    }
  })
}
