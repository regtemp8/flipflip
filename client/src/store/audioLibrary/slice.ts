import { createSlice, type PayloadAction } from '@reduxjs/toolkit'

interface AudioLibraryState {
  error: boolean
  loadingMetadata: boolean
  loadingSources: boolean
  drawerOpen: boolean
  selectedTags: string[]
  openMenu?: string
  playlistID?: number
  importURL?: string
}

export const initialAudioLibraryState: AudioLibraryState = {
  error: false,
  loadingMetadata: false,
  loadingSources: false,
  drawerOpen: false,
  selectedTags: []
}

export const audioLibrarySlice = createSlice({
  name: 'audioLibrary',
  // `createSlice` will infer the state type from the `initialState` argument
  initialState: initialAudioLibraryState,
  reducers: {
    setError: (state, action: PayloadAction<boolean>) => {
      state.error = action.payload
    },
    setLoadingMetadata: (state, action: PayloadAction<boolean>) => {
      state.loadingMetadata = action.payload
    },
    setLoadingSources: (state, action: PayloadAction<boolean>) => {
      state.loadingSources = action.payload
    },
    setOpenMenu: (state, action: PayloadAction<string | undefined>) => {
      state.openMenu = action.payload
    },
    setPlaylistID: (state, action: PayloadAction<number | undefined>) => {
      state.playlistID = action.payload
    },
    setDrawerOpen: (state, action: PayloadAction<boolean>) => {
      state.drawerOpen = action.payload
    },
    setImportURL: (state, action: PayloadAction<string | undefined>) => {
      state.importURL = action.payload
    },
    setSelectedTags: (state, action: PayloadAction<string[]>) => {
      state.selectedTags = action.payload
    },
    setAudioLibraryPlaylistID: (state, action: PayloadAction<string>) => {
      state.playlistID = Number(action.payload)
    }
  }
})

export const {
  setError,
  setLoadingMetadata,
  setLoadingSources,
  setOpenMenu,
  setPlaylistID,
  setDrawerOpen,
  setImportURL,
  setSelectedTags,
  setAudioLibraryPlaylistID
} = audioLibrarySlice.actions

export default audioLibrarySlice.reducer
