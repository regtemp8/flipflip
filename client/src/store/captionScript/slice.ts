import { createSlice, type PayloadAction } from '@reduxjs/toolkit'
import type CaptionScript from './CaptionScript'
import {
  type EntryState,
  type EntryUpdate,
  setEntrySlice,
  setEntry,
  updateEntry
} from '../EntryState'
import { type FontSettingsType } from './FontSettings'
import { getCaptionScript, getCaptionScriptFontSettings } from './selectors'

export interface FontSettingsUpdate<T> extends EntryUpdate<T> {
  type: FontSettingsType
}

export const initialCaptionScriptState: EntryState<CaptionScript> = {
  name: 'captionScriptSlice',
  nextID: 1,
  entries: {}
}
export const captionScriptSlice = createSlice({
  name: 'captionScripts',
  // `createSlice` will infer the state type from the `initialState` argument
  initialState: initialCaptionScriptState,
  reducers: {
    setCaptionScriptSlice: (
      state,
      action: PayloadAction<EntryState<CaptionScript>>
    ) => {
      setEntrySlice(state, action.payload)
    },
    setCaptionScript: (state, action: PayloadAction<CaptionScript>) => {
      setEntry(state, action.payload)
    },
    updateCaptionScript: (state, action: PayloadAction<CaptionScript>) => {
      updateEntry(state, action.payload)
    },
    setCaptionScriptScript: (
      state,
      action: PayloadAction<EntryUpdate<string>>
    ) => {
      getCaptionScript(state, action.payload.id).script = action.payload.value
    },
    setCaptionScriptURL: (
      state,
      action: PayloadAction<EntryUpdate<string>>
    ) => {
      getCaptionScript(state, action.payload.id).url = action.payload.value
    },
    setCaptionScriptStopAtEnd: (
      state,
      action: PayloadAction<EntryUpdate<boolean>>
    ) => {
      getCaptionScript(state, action.payload.id).stopAtEnd =
        action.payload.value
    },
    setCaptionScriptNextSceneAtEnd: (
      state,
      action: PayloadAction<EntryUpdate<boolean>>
    ) => {
      getCaptionScript(state, action.payload.id).nextSceneAtEnd =
        action.payload.value
    },
    setCaptionScriptSyncWithAudio: (
      state,
      action: PayloadAction<EntryUpdate<boolean>>
    ) => {
      getCaptionScript(state, action.payload.id).syncWithAudio =
        action.payload.value
    },
    setCaptionScriptOpacity: (
      state,
      action: PayloadAction<EntryUpdate<number>>
    ) => {
      getCaptionScript(state, action.payload.id).opacity = action.payload.value
    },
    setCaptionScriptFontSettingsBorder: (
      state,
      action: PayloadAction<FontSettingsUpdate<boolean>>
    ) => {
      getCaptionScriptFontSettings(
        state,
        action.payload.id,
        action.payload.type
      ).border = action.payload.value
    },
    setCaptionScriptFontSettingsColor: (
      state,
      action: PayloadAction<FontSettingsUpdate<string>>
    ) => {
      getCaptionScriptFontSettings(
        state,
        action.payload.id,
        action.payload.type
      ).color = action.payload.value
    },
    setCaptionScriptFontSettingsBorderColor: (
      state,
      action: PayloadAction<FontSettingsUpdate<string>>
    ) => {
      getCaptionScriptFontSettings(
        state,
        action.payload.id,
        action.payload.type
      ).borderColor = action.payload.value
    },
    setCaptionScriptFontSettingsFontFamily: (
      state,
      action: PayloadAction<FontSettingsUpdate<string>>
    ) => {
      getCaptionScriptFontSettings(
        state,
        action.payload.id,
        action.payload.type
      ).fontFamily = action.payload.value
    },
    setCaptionScriptFontSettingsFontSize: (
      state,
      action: PayloadAction<FontSettingsUpdate<number>>
    ) => {
      getCaptionScriptFontSettings(
        state,
        action.payload.id,
        action.payload.type
      ).fontSize = action.payload.value
    },
    setCaptionScriptFontSettingsBorderPx: (
      state,
      action: PayloadAction<FontSettingsUpdate<number>>
    ) => {
      getCaptionScriptFontSettings(
        state,
        action.payload.id,
        action.payload.type
      ).borderpx = action.payload.value
    },
    setCaptionScriptTags: (
      state,
      action: PayloadAction<EntryUpdate<number[]>>
    ) => {
      state.entries[action.payload.id].tags = action.payload.value
    },
    setCaptionScriptToggleTag: (
      state,
      action: PayloadAction<EntryUpdate<number>>
    ) => {
      const script = state.entries[action.payload.id]
      const index = script.tags.indexOf(action.payload.value)
      if (index === -1) {
        script.tags.push(action.payload.value)
      } else {
        script.tags.splice(index, 1)
      }
    },
    setCaptionScriptMarked: (
      state,
      action: PayloadAction<EntryUpdate<boolean>>
    ) => {
      getCaptionScript(state, action.payload.id).marked = action.payload.value
    }
  }
})

export const {
  setCaptionScriptSlice,
  setCaptionScript,
  updateCaptionScript,
  setCaptionScriptScript,
  setCaptionScriptURL,
  setCaptionScriptStopAtEnd,
  setCaptionScriptNextSceneAtEnd,
  setCaptionScriptSyncWithAudio,
  setCaptionScriptOpacity,
  setCaptionScriptFontSettingsBorder,
  setCaptionScriptFontSettingsColor,
  setCaptionScriptFontSettingsBorderColor,
  setCaptionScriptFontSettingsFontFamily,
  setCaptionScriptFontSettingsFontSize,
  setCaptionScriptFontSettingsBorderPx,
  setCaptionScriptTags,
  setCaptionScriptToggleTag,
  setCaptionScriptMarked
} = captionScriptSlice.actions

export default captionScriptSlice.reducer
