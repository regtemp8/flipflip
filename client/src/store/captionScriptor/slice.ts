import { createSlice, type PayloadAction } from '@reduxjs/toolkit'

interface CaptionScriptorState {
  sceneID: number
  captionScriptID: number
  sceneScripts: string[]
}

export const initialCaptionScriptorState: CaptionScriptorState = {
  sceneID: 0,
  captionScriptID: -1,
  sceneScripts: []
}
export const captionScriptorSlice = createSlice({
  name: 'captionScriptor',
  // `createSlice` will infer the state type from the `initialState` argument
  initialState: initialCaptionScriptorState,
  reducers: {
    setCaptionScriptorSceneID: (state, action: PayloadAction<number>) => {
      state.sceneID = action.payload
    },
    setCaptionScriptorSceneScripts: (
      state,
      action: PayloadAction<string[]>
    ) => {
      state.sceneScripts = action.payload
    },
    setCaptionScriptorCaptionScriptID: (
      state,
      action: PayloadAction<number | undefined>
    ) => {
      state.captionScriptID = action.payload ?? -1
    }
  }
})

export const {
  setCaptionScriptorSceneID,
  setCaptionScriptorSceneScripts,
  setCaptionScriptorCaptionScriptID
} = captionScriptorSlice.actions

export default captionScriptorSlice.reducer
