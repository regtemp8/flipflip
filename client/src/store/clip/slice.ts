import { createSlice, type PayloadAction } from '@reduxjs/toolkit'
import type Clip from './Clip'
import {
  type EntryState,
  type EntryUpdate,
  setEntrySlice,
  setEntry
} from '../EntryState'

export const initialClipState: EntryState<Clip> = {
  name: 'clipSlice',
  nextID: 1,
  entries: {}
}

export default function createClipReducer(clipState?: EntryState<Clip>) {
  return createClipSlice(clipState).reducer
}

function createClipSlice(clipState?: EntryState<Clip>) {
  const initialState = clipState ?? initialClipState
  return createSlice({
    name: 'clips',
    // `createSlice` will infer the state type from the `initialState` argument
    initialState,
    reducers: {
      setClipSlice: (state, action: PayloadAction<EntryState<Clip>>) => {
        setEntrySlice(state, action.payload)
      },
      setClip: (state, action: PayloadAction<Clip>) => {
        setEntry(state, action.payload)
      },
      setClips: (state, action: PayloadAction<Clip[]>) => {
        action.payload.forEach((s) => setEntry(state, s))
      },
      setClipToggleTag: (state, action: PayloadAction<EntryUpdate<number>>) => {
        const clip = state.entries[action.payload.id]
        const index = clip.tags.indexOf(action.payload.value)
        if (index === -1) {
          clip.tags.push(action.payload.value)
        } else {
          clip.tags.splice(index, 1)
        }
      },
      setClipVolume: (
        state,
        action: PayloadAction<EntryUpdate<number | undefined>>
      ) => {
        state.entries[action.payload.id].volume = action.payload.value
      },
      setClipTags: (state, action: PayloadAction<EntryUpdate<number[]>>) => {
        state.entries[action.payload.id].tags = action.payload.value
      },
      setClipStartEnd: (
        state,
        action: PayloadAction<EntryUpdate<number[]>>
      ) => {
        const clip = state.entries[action.payload.id]
        clip.start = action.payload.value[0]
        clip.end = action.payload.value[1]
      }
    }
  })
}

export const {
  setClipSlice,
  setClip,
  setClips,
  setClipToggleTag,
  setClipTags,
  setClipVolume,
  setClipStartEnd
} = createClipSlice().actions
