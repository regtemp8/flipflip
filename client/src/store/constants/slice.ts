import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import { initialSystemConstants } from './SystemConstants'
import { SystemConstants, copy } from 'flipflip-common'

export default function createConstantsReducer(constants?: SystemConstants) {
  return createConstantsSlice(constants).reducer
}

function createConstantsSlice(constants?: SystemConstants) {
  let initialState = copy<SystemConstants>(initialSystemConstants)
  if (constants != null) {
    initialState = Object.assign(initialState, constants)
  }

  return createSlice({
    name: 'constants',
    // `createSlice` will infer the state type from the `initialState` argument
    initialState,
    reducers: {
      setConstants: (state, action: PayloadAction<SystemConstants>) => {
        return action.payload
      }
    }
  })
}

export const { setConstants } = createConstantsSlice().actions
