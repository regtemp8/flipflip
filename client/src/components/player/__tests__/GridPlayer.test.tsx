import React from 'react'
import { describe, it, expect } from '@jest/globals'
import renderer from 'react-test-renderer'
import GridPlayer from '../GridPlayer'
import TestProvider from '../../../util/TestProvider'
import store from '../../../store/store'
import { setSceneGrid } from '../../../store/sceneGrid/slice'
import { newSceneGrid } from '../../../store/sceneGrid/SceneGrid'
import {
  PlayerOverlayState,
  PlayersState,
  setPlayersState
} from '../../../store/player/slice'
import { setScene } from '../../../store/scene/slice'
import { newScene } from '../../../store/scene/Scene'
import { newSceneGridCell } from '../../../store/sceneGrid/SceneGridCell'

jest.mock('../Player', () => 'Player')

// TODO create functional tests instead of snapshots
describe('GridPlayer', () => {
  it('should match snapshot', () => {
    const sceneID = 3
    const overlayState: PlayerOverlayState = {
      id: 1,
      opacity: 1,
      sceneID,
      isGrid: true,
      grid: [['grid']],
      loaded: false
    }
    const playersState: PlayersState = {
      root: {
        sceneID: 1,
        nextSceneID: 0,
        overlays: [overlayState],
        firstImageLoaded: false,
        mainLoaded: false,
        isEmpty: false,
        hasStarted: false
      },
      grid: {
        sceneID: 1,
        nextSceneID: 0,
        overlays: [],
        firstImageLoaded: false,
        mainLoaded: false,
        isEmpty: false,
        hasStarted: false
      }
    }

    store.dispatch(setScene(newScene({ id: 1 })))
    store.dispatch(
      setSceneGrid(
        newSceneGrid({
          id: sceneID,
          grid: [[newSceneGridCell({ sceneID: 1 })]]
        })
      )
    )
    store.dispatch(setPlayersState(playersState))
    const component = renderer.create(
      <TestProvider store={store}>
        <GridPlayer
          parentUUID="root"
          overlayIndex={0}
          gridConfig={overlayState}
        />
      </TestProvider>
    )

    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
