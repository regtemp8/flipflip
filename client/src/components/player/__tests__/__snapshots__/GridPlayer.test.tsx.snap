// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GridPlayer should match snapshot 1`] = `
<div
  className="GridPlayer-root-1"
>
  <div
    className="GridPlayer-hoverBar-4"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  />
  <header
    className="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionAbsolute GridPlayer-appBar-2 css-iu2t21-MuiPaper-root-MuiAppBar-root"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      className="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
    >
      <button
        aria-label="Back"
        aria-labelledby={null}
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart MuiIconButton-sizeLarge css-184mzhy-MuiButtonBase-root-MuiIconButton-root"
        data-mui-internal-clone-element={true}
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ArrowBackIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
          />
        </svg>
        <span
          className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <div
        className="GridPlayer-fill-10"
      />
      <h1
        className="MuiTypography-root MuiTypography-h4 MuiTypography-noWrap GridPlayer-title-5 css-1r3v6n7-MuiTypography-root"
      />
      <div
        className="GridPlayer-fill-10"
      />
      <button
        aria-label="FullScreen"
        aria-labelledby={null}
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart MuiIconButton-sizeLarge css-184mzhy-MuiButtonBase-root-MuiIconButton-root"
        data-mui-internal-clone-element={true}
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge css-tzssek-MuiSvgIcon-root"
          data-testid="FullscreenIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 14H5v5h5v-2H7v-3zm-2-4h2V7h3V5H5v5zm12 7h-3v2h5v-5h-2v3zM14 5v2h3v3h2V5h-5z"
          />
        </svg>
        <span
          className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </header>
  <main
    className="GridPlayer-content-6"
  >
    <div />
    <div
      className="MuiContainer-root GridPlayer-container-7 css-ur2jdm-MuiContainer-root"
    >
      <div
        className="GridPlayer-grid-8"
        style={
          {
            "gridTemplateColumns": "100% ",
            "gridTemplateRows": "100% ",
          }
        }
      >
        <div
          className="GridPlayer-gridCell-9 GridPlayer-hidden-11"
        />
      </div>
    </div>
  </main>
</div>
`;
